src/lib/shell/commands/WiFi.cpp

```diff
@@ -104,13 +109,69 @@ static CHIP_ERROR WiFiModeHandler(int argc, char ** argv)
 
 static CHIP_ERROR WiFiConnectHandler(int argc, char ** argv)
 {
-    if (argc != 2)
+    CHIP_ERROR error = CHIP_NO_ERROR;
+    uint8_t networkIndex;
+    MutableCharSpan debugText;
+    ByteSpan ssidSpan;
+    ByteSpan passwordSpan;
+
+    VerifyOrReturnError(GetWiFiDriver() != nullptr, error = CHIP_ERROR_NOT_IMPLEMENTED);
+
+    /* Command accepts running with SSID and password as parameters */
+    VerifyOrReturnError((argc == 2), error = CHIP_ERROR_INVALID_ARGUMENT);
+
+    ssidSpan     = ByteSpan(reinterpret_cast<const uint8_t *>(argv[0]), strlen(argv[0]));
+    passwordSpan = ByteSpan(reinterpret_cast<const uint8_t *>(argv[1]), strlen(argv[1]));
+
+    if (IsSpanUsable(ssidSpan) && IsSpanUsable(passwordSpan))
     {
-        return CHIP_ERROR_INVALID_ARGUMENT;
+        ChipLogProgress(DeviceLayer, "[Shell] Adding/Updating network %s", ssidSpan.data());
+
+        /* AddOrUpdateNetwork() checks ssid length and password length. The network info is not persistent. */
+        GetWiFiDriver()->AddOrUpdateNetwork(ssidSpan, passwordSpan, debugText, networkIndex);
+
+        ChipLogProgress(DeviceLayer, "[Shell] Progress: Connecting to network");
+        /* Connection event will be returned in OnWiFiConnectivityChange from DeviceCallbacks.cpp */
+        GetWiFiDriver()->ConnectNetwork(ssidSpan, nullptr);
+    }
+    else
+    {
+        ChipLogError(DeviceLayer, "[Shell] Failed to add network credentials!");
+        error = CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    return error;
+}
+
+static CHIP_ERROR WiFiDisconnectHandler(int argc, char ** argv)
+{
+    VerifyOrReturnError((argc == 0), CHIP_ERROR_INVALID_ARGUMENT);
+
+    return ConnectivityMgr().DisconnectNetwork();
+}
+
+static CHIP_ERROR WiFiScanHandler(int argc, char ** argv)
+{
+    ByteSpan ssidSpan;
+
+    VerifyOrReturnError(GetWiFiDriver() != nullptr, CHIP_ERROR_NOT_IMPLEMENTED);
+
+    /* Command accepts running either without arguments or either with ssid */
+    /* Running with argument ssid shall restrict the scan to the interested ssid */
+    VerifyOrReturnError((argc == 0) || (argc == 1), CHIP_ERROR_INVALID_ARGUMENT);
+
+    if (argc == 0)
+    {
+        ssidSpan = ByteSpan();
+    }
+    else if (argc == 1)
+    {
+        ssidSpan = ByteSpan(reinterpret_cast<const uint8_t *>(argv[0]), strlen(argv[0]));
     }
 
-    // TODO:Provision WiFi using WirelessDriver
-    return CHIP_ERROR_NOT_IMPLEMENTED;
+    GetWiFiDriver()->ScanNetworks(ssidSpan, nullptr);
```

Comment:
Don't you have to provide the callback to have any results printed?

Reply:
I removed scan function for now, I don't know how to handle it across all vendors at this point. On my side, I tried it without the parameter and I modified the ScanNetworks function to not take into account the callback set here, which is wrong, but it worked because I had a callback registered from the platform wifi driver. I will see if I find any solution and come back with another PR.
Thank you very much for your input, @Damian-Nordic .

Reply:
I guess shell command can provide its own callback but it can be done later.